docs: add overview of session recording

zmb3 opened the pull request:
After speaking with a few people I found that there isn't a good overview of session recording in our docs, only various bits and pieces in different sections.

This aims to address that by adding an overview that speaks to how session recording works in general and the various configuration options that apply.

zmb3 committed:
docs: add overview of session recording

Commit:
add new doc to config

Commit:
formatting

Commit:
Minor copy edits

zmb3 commented on docs/pages/reference/session-recording.mdx:
```diff
@@ -0,0 +1,141 @@
+
+This results in 4 possible sesssion recording configurations:
+
+- `node-sync`: synchronous recording performed by the SSH node
+- `node`: asynchronous recording performed by the SSH node
```
> We're replacing "SSH" with "Server Access", right?

That would be news to me. Though I don't see a harm in sticking with SSH here since it's the most specific.

> This is true for Server Access sessions, but what about others?

Well, that's the thing, SSH is unique in that Teleport software is running at both ends of the connection (Teleport is the SSH server you're connecting to, and Teleport is the SSH client)

For desktop access (as mentioned elsewhere in this PR) there is not a distinction between "node" and "proxy" - the recording is always done on the windows_desktop_service, which is neither a node or proxy. 

Kube and DB access are similar - the Kube cluster or DB that you're connecting to is not Teleport software, it's other software that's totally unaware of Teleport, just like a desktop with RDP.

alexfornuto replied:
> > We're replacing "SSH" with "Server Access", right?
> 
> That would be news to me. Though I don't see a harm in sticking with SSH here since it's the most specific.

Maybe I have that backwards.

alexfornuto replied:
> Kube and DB access are similar - the Kube cluster or DB that you're connecting to is not Teleport software, it's other software that's totally unaware of Teleport, just like a desktop with RDP.

You mean similar to Desktop, where the recoding is done at the node running the service, right?

alexfornuto replied:
So the more accurate statement is that session recording for DB and Kube occurs at the host running the Teleport Service for Database and Kubernetes access?

zmb3 replied:
Right. Kube, DB, Desktop (and even App Access) are all alike in this regard, while SSH is a bit unique, which is why I mentioned SSH specifically in the original bullet point that started this thread.

alexfornuto commented on docs/pages/reference/session-recording.mdx:
```diff
@@ -0,0 +1,141 @@
+their own recording configurations.
+
+<Admonition type="note" title="Desktop Access">
+Desktop Sessions are always recorded at Teleport's Windows Desktop Service
+since there is no Teleport software running on the "node" (desktop).
+
+For this reason, Teleport Desktop Access treats `node` and `proxy` identically
+(perform synchronous desktop recording) and `node-sync` and `proxy-sync` identically
+(perform asynchronous desktop recording).
```
Suggested to address the discussion thread above.

```suggestion
Desktop, Database, and Kubernetes Sessions are always recorded at the host running
the Teleport service for that session type, since there is no Teleport software
running on the "node" (desktop, database, or Kubernetes cluster).

For this reason, Teleport Desktop, Database, and Kubernetes Access treats `node`
and `proxy` identically (perform synchronous recording) and `node-sync` and
`proxy-sync` identically (perform asynchronous recording).
```

alexfornuto replied:
Adjusted after discussion.

webvictim commented on docs/pages/reference/session-recording.mdx:
```diff
@@ -0,0 +1,141 @@
+with a session. Each event is an encoded [Protocol Buffer](https://developers.google.com/protocol-buffers)
+and the complete session is compressed with gzip before being written to the storage backend.
+
+These recordings have a `.tar` extension for backwards compatibility reason, but it
+should be noted that they are not TAR archives and cannot be read using the `tar` utility.
```
Someone told me once that it stands for **T**eleport **Ar**chive 😁

webvictim reviewed (commented):
Could we maybe add some detail here about alternative failure scenarios - what happens in sync/async modes when the Teleport process which is handling the session is killed, restarted or dies? 

My understanding is that there's some kind of background process that looks at sessions which haven't completed and _eventually_ completes them - but how does this work, and what expectation is there on how long it should take to "finalise" a log for a session that failed to upload?

Commit:
adjust note for all non-node services

Commit:
add links

alexfornuto reviewed (approved):
This would be a great addition to the docs. I agree with @webvictim's suggestion but lack the subject matter expertise to provide the additional information myself.

IMO the author could add the requested content in this PR, or we could create a new issue to update this after publication, to avoid holding this content back.

Commit:
validate local links

zmb3 committed:
Add section on upload completer

ptgott commented on docs/pages/reference/session-recording.mdx:
```diff
@@ -0,0 +1,162 @@
+
+Teleport supports two different recording configurations: synchronous and
+asynchronous recording. Additionally, Teleport administrators can configure
+where the recording takes place: at the node or at the proxy.
```
We should make "node" and "proxy" are consistent with the rest of the docs (in terms of usage and capitalization), especially the Core Concepts page (https://goteleport.com/docs/core-concepts/). Does "proxy" refer to "Proxy Service" here? Also, does "node" refer to an SSH Service instance?

zmb3 replied:
Kind of. The use of these terms is somewhat intentional, as:

1. the config setting for these are literally `node` or `proxy`.
2. "proxy" typically means the "Proxy Service" but may also refer to the kube service or the windows desktop service, so the more generic proxy makes sense.
3. yes, 'node" refers to the SSH service

ptgott commented on docs/pages/reference/session-recording.mdx:
```diff
@@ -0,0 +1,162 @@
+description: An overview of Teleport's session recording and its configuration
+---
+
+This page provides an overview of how Teleport records sessions and the various
```
We might want to expand on what session recording is earlier in the page. What is a session? What does it mean to record it?

zmb3 replied:
Perhaps I'm blinded by my own experience, as it's pretty obvious to me. Do you have any suggestions for what we might say in an expanded explanation?

zmb3 committed:
fix lint

ptgott commented on docs/pages/reference/session-recording.mdx:
```diff
@@ -0,0 +1,162 @@
+with their own recording configurations.
+
+<Admonition type="note">
+Windows, Database, and Kubernetes Sessions are always recorded at the host running
```
```suggestion
Windows, Database, and Kubernetes sessions are always recorded at the host running
```

ptgott reviewed (commented):
I'll have time to give this a full review later, but here are some initial comments.

Commit:
Update docs/pages/reference/session-recording.mdx

Co-authored-by: Paul Gottschling <paul.gottschling@goteleport.com>

ptgott commented on docs/pages/reference/session-recording.mdx:
```diff
@@ -0,0 +1,162 @@
+
+Session recordings are stored in Teleport's audit sessions backend, which is
+specified by the `audit_sessions_uri` field in the `teleport.yaml` [configuration file](config.mdx).
+Teleport current supports the following session storage backends:
```
Should we list the values you'd put in the configuration file below?

zmb3 replied:
Personally, I don't think so. There's quite a few permutations and it's already covered in detail in the config reference. This page was more meant to be about how it works rather than how to configure it.

ptgott commented on docs/pages/reference/session-recording.mdx:
```diff
@@ -0,0 +1,162 @@
+  Suitable for production deployments.
+- GCS: stores recordings in Google Cloud Storage. Suitable for production deployments.
+
+The Teleport Auth server is the only component that writes directly to the storage backend.
```
```suggestion
The Teleport Auth Service is the only component that writes directly to the storage backend.
```

ptgott commented on docs/pages/reference/session-recording.mdx:
```diff
@@ -0,0 +1,162 @@
+The Teleport Auth server is the only component that writes directly to the storage backend.
+
+Note that the session storage backend is distinct from Teleport's cluster state or
+audit log storage, which supports a different set of services (SQLite, DynamoDB, Firestore, etc)
```
```suggestion
audit log storage, which supports a different set of services (SQLite, DynamoDB, Firestore, etc).
```

ptgott commented on docs/pages/reference/session-recording.mdx:
```diff
@@ -0,0 +1,162 @@
+and audit log are stored in separate backends, both need to be operational in
+order to play recordings.
+
+## Upload Completer
```
```suggestion
## Upload completer
```
Since we're lowercasing "upload completer"

ptgott commented on docs/pages/reference/session-recording.mdx:
```diff
@@ -0,0 +1,162 @@
+responsibility of the Auth Server's upload completer to detect the abandoned upload and
+complete it.
+
+## Related Reading
```
```suggestion
## Related reading
```

zmb3 commented on docs/pages/reference/session-recording.mdx:
```diff
@@ -0,0 +1,162 @@
+This page provides an overview of how Teleport records sessions and the various
+configuration options that apply to session recordings.
+
+Teleport supports two different recording configurations: synchronous and
```
Ack. I'm being very careful to say "configurations" and not "recording modes" here, because unfortunately some recent work has stolen the term "recording modes" and I don't want to go overboard in this doc.

zmb3 commented on docs/pages/reference/session-recording.mdx:
```diff
@@ -0,0 +1,162 @@
+
+By default, Teleport performs recording at the SSH node. This is because Teleport's
+Proxy Server cannot see the SSH traffic to the node, it is encrypted end-to-end
+(from the SSH client to the SSH server).
```
SSH client can be `tsh` or OpenSSH's `ssh` client.

SSH server can be Teleport's SSH service or OpenSSH's sshd.

zmb3 commented on docs/pages/reference/session-recording.mdx:
```diff
@@ -0,0 +1,162 @@
+with their own recording configurations.
+
+<Admonition type="note">
+Windows, Database, and Kubernetes sessions are always recorded at the host running
```
What this is trying to say is that the node/proxy distinction only matters for SSH sessions. For desktop or kubernetes sessions, `node` and `proxy` are equivalent, and `node-sync` and `proxy-sync` are equivalent.

If you have suggestions for a better way to phrase this I'm all ears!

zmb3 committed:
Address Paul's feedback

zmb3 commented:
@ptgott I took a stab and addressed most of your feedback. Let me know what you think about the remaining items. I left anything that wasn't clear cut unresolved.

zmb3 committed:
Add info on DB and app recordings too

Most of the mechanics work the same way as for other sessions.
The difference is these sessions have a more limited "playback"
ability since they're just a group of audit metadata.

ptgott commented on docs/pages/reference/session-recording.mdx:
```diff
@@ -0,0 +1,179 @@
+---
+
+This page provides an overview of how Teleport records sessions and the various
+configuration options that apply to session recordings. The content that is
+captured depends on the type of the session.
```
```suggestion
configuration options that apply to session recordings. 

## What session recordings capture

The content session recordings capture depends on the type of the session.
```
Since this information is pretty granular, I thought it would make sense to separate it from the intro paragraph.

ptgott commented on docs/pages/reference/session-recording.mdx:
```diff
@@ -0,0 +1,179 @@
+- Database sessions: Teleport captures a stream of audit events
+  related to the database being accessed.
+
+Teleport supports two different recording configurations: synchronous and
```
```suggestion
## Session recording configurations

Teleport supports two different recording configurations: synchronous and
```
Suggested because this is a different topic area than the earlier breakdown of what session recording captures

ptgott commented on docs/pages/reference/session-recording.mdx:
```diff
@@ -0,0 +1,179 @@
+`proxy-sync` identically (perform asynchronous recording).
+</Admonition>
+
+## Record at Node
```
```suggestion
### Record at Node
```
To place this in the "Session recording configurations" section I suggested earlier

ptgott commented on docs/pages/reference/session-recording.mdx:
```diff
@@ -0,0 +1,179 @@
+Proxy Server cannot see the SSH traffic to the node, it is encrypted end-to-end
+(from the SSH client to the SSH server).
+
+## Record at Proxy
```
```suggestion
### Record at Proxy
```

ptgott commented on docs/pages/reference/session-recording.mdx:
```diff
@@ -0,0 +1,179 @@
+
+In some cases, it is desirable to perform the recording at Teleport's Proxy Server.
+For example, sessions to non-Teleport servers (like OpenSSH's `sshd`) can only be
+recorded at the proxy. This is referred to as *Recording Proxy Mode*.
```
```suggestion
recorded at a Proxy Service instance. This is referred to as *Recording Proxy Mode*.
```

ptgott commented on docs/pages/reference/session-recording.mdx:
```diff
@@ -0,0 +1,179 @@
+it must be properly secured and is a higher value target for an attacker than a proxy
+that cannot decrypt the data flowing through it.
+
+Additionally, the credentials that the proxy uses to decrypt the SSH connection are
```
```suggestion
Additionally, the credentials that the Proxy Server uses to decrypt the SSH connection are
```

ptgott commented on docs/pages/reference/session-recording.mdx:
```diff
@@ -0,0 +1,179 @@
+provided via SSH Agent Forwarding, so Agent Forwarding must be enabled to record at
+the Proxy Server.
+
+## Synchronous recording
```
```suggestion
### Synchronous recording
```

ptgott commented on docs/pages/reference/session-recording.mdx:
```diff
@@ -0,0 +1,179 @@
+## Synchronous recording
+
+When synchronous recording is enabled, the Teleport component doing the recording
+(which may be the node or the proxy depending on your configuration) submits each
```
```suggestion
(which may be the Node or the Proxy Service instance depending on your configuration) submits each
```

ptgott commented on docs/pages/reference/session-recording.mdx:
```diff
@@ -0,0 +1,179 @@
+In synchronous recording modes, the Auth Server receives a stream of recording
+events and is responsible for assembling them into the final artifact and uploading
+it to the storage backend. Since data is streamed directly to the Auth Server,
+Teleport administrators don't need to be concerned with disk space on their nodes
+and proxies, as no recording data is written to those disks.
```
```suggestion
Teleport administrators don't need to be concerned with disk space on their Nodes
and Proxy Service instances, as no recording data is written to those disks.
```

ptgott commented on docs/pages/reference/session-recording.mdx:
```diff
@@ -0,0 +1,179 @@
+Teleport administrators don't need to be concerned with disk space on their nodes
+and proxies, as no recording data is written to those disks.
+
+## Asynchronous recording
```
```suggestion
### Asynchronous recording
```

ptgott reviewed (approved):
Left suggestions and approved so you're not blocked

zmb3 commented on docs/config.json:
```diff
@@ -1103,6 +1103,10 @@
           "title": "Audit Events and Records",
           "slug": "/reference/audit/"
         },
+        {
```
Yeah, that makes a lot of sense! I'll move it.

zmb3 committed:
Another review pass

zmb3 committed:
Move session recording overview to architecture section

zmb3 committed:
Fix broken links
